go/storage/client/client.go

```diff
@@ -305,9 +305,33 @@ func (b *storageClientBackend) readWithClient(
 		// https://github.com/oasisprotocol/oasis-core/issues/1815.
 		rng := rand.New(mathrand.New(cryptorand.Reader))
 
+		// If a storage node priority hint is set, prioritize overlapping nodes.
+		priorityNodes := api.NodePriorityHintFromContext(ctx)
+		nodeOrder := make([]int, n)
+		prioritySlots := len(priorityNodes)
+		for i, c := range conns {
+			// If in the priority list, add to slot based on priority list order.
+			if prioritySlots > 0 {
+				prioritySlot := -1
+				for slot, nodeID := range priorityNodes {
+					if c.Node.ID.Equal(nodeID) {
+						prioritySlot = slot
+						break
+					}
+				}
+				if prioritySlot > -1 {
+					prioritySlots--
+					nodeOrder[prioritySlot] = i
+					continue
+				}
+			}
+			// Otherwise add to some random slot after the priority slots.
+			nodeOrder[rng.Intn(n-len(priorityNodes))+len(priorityNodes)] = i
```

Comment:
I don't think this last part is correct as it can insert into the same random slot multiple times.

Reply:
Yeah it's not.